batcher: keep blocks, channels and frames in strict order & simplify reorg handling

geoknee opened the pull request:
Closes https://github.com/ethereum-optimism/optimism/issues/12123

Design doc https://www.notion.so/oplabs/op-batcher-re-architecture-114f153ee162803d943ff4628ab6578f?pvs=4

Changes (at the reviewer's request I could split this PR up):
1. Use the `op-service/queue` type in a couple more places in the batcher
2. Introduce cursors and reduce the amount of queue operations so that we can be sure things stay in order
3. Simplify reorg handling (it was causing the channel queue to be spliced up which is contrary to the aims of (2))

---
From the issue: 

_We should check_

- [x] normal operation always enqueues blocks inside in order {This was already the case, but now we only ever enqueue and dequeue, never prepend blocks). 
- [x] keeps channels in the channel queue in order (We now only ever enqueue, dequeue or delete from the back of the queue in a way which preserves the order -- by rewinding the block cursor)
- [x] sends transactions in order (This was already true, the txmgr has a queue which we add txs to in order)
- [x] also frames within blob transactions in order (This was already true, frames are appended to the transaction in order)
- [x] reorg case (We were already clearing out the batcher's state and starting again when detecting this. Continuing to do that, only now we don't bother finishing any tx submission work before doing so)
- [x] re-queueing case when a channel fails to submit in time -> need to preserver order across channels (also future pending) (When this happens, we now clear out the channel queue and rewind the block cursor to point at the first block of the timed out channel). 
- [x] re-queueing case when da-type changes (This was already OK, but now simplified to make it clear we only ever operate with a single channel, the current channel). Again we rewind the block cursor so that the blocks can be reprocessed).

geoknee committed:
use a queue.Queue for channelBuilder.frames

geoknee committed:
remove pop and push terminology

geoknee committed:
proliferate queue.Queue type

geoknee committed:
simplify requeue method

geoknee committed:
undo changes to submodule

geoknee committed:
sketch out new arch

https://www.notion.so/oplabs/op-batcher-re-architecture-114f153ee162803d943ff4628ab6578f

geoknee committed:
add TODO

geoknee committed:
add channelManager.pruneSafeBlocks method and integrate into main loop

geoknee committed:
fix frameCursor semantics

geoknee committed:
fixup tests

geoknee committed:
avoid Rewind() in tests

geoknee committed:
only rewind cursor in rewind (never move it forward)

geoknee committed:
fix assertions

geoknee committed:
prune channels whose blocks are now safe

geoknee committed:
handle case when rewinding a channel with no blocks

this is strange, I don't think we should expect channels with frames but no blocks...

geoknee committed:
add clarification

geoknee committed:
implement channelManager.pendinBlocks() method

geoknee committed:
fix pruning logic

geoknee committed:
simplify pruneChannels

geoknee committed:
simplify pruneSafeBlocks

geoknee committed:
add unit tests for pruneSafeBlocks

geoknee committed:
fix pruneSafeBlocks to avoid underflow

geoknee committed:
improve test

geoknee committed:
add unit tests for pruneChannels

geoknee committed:
introduce handleChannelTimeout

and simplify channel.TxConfirmed API

geoknee committed:
factor out channelManager.rewindToBlockWithHash

geoknee committed:
change test expectation

geoknee committed:
do more pruning in test

geoknee committed:
Replace "clean shutdown" behaviour with waitNodeSync()

Instead of optimizing for a clean shutdown (which we couldn't guarantee anyway), this change optimizes for code simplicity.

This change also helps us restrict the amount of code which mutates the channelQueue (removePendingChannel was doing removal of channels at arbitrary positions in the queue).

The downside is that we may end up needlessly resubmitting some data after the reset.

Reorgs are rare, so it makes sense to optimize for correctness rather than DA costs here.

geoknee committed:
Add readme and architecture diagram

geoknee committed:
don't panic when there is a safe chain reorg

geoknee committed:
fix test

geoknee committed:
readability improvements

geoknee committed:
only clear state after waiting for node to sync

geoknee committed:
resize image

geoknee committed:
tweak readme

vdamle commented on op-batcher/batcher/channel.go:
```diff
@@ -67,39 +67,36 @@ func (s *channel) TxFailed(id string) {
 	s.confirmedTxUpdated = true
 	s.channelBuilder.FramePublished(inclusionBlock.Number)
 
+	if s.isFullySubmitted() {
```
Intuitively, it makes sense that we'd check whether the channel was fully submitted before checking if it timed out. Out of curiosity, were we hitting any case where the channel would be marked as timed out, erroneously (and hence you've shifted the check earlier)?

geoknee replied:
The motivation behind this change is that if I wanted to make sure the `s.isFullySubmitted` check ran unconditionally. Previously, it would not run if the channel had timed out because of the `return` on L93. Does that make sense to you?

geoknee replied:
Let's double check how channels are cleaned up / garbage collected, because the surrounding code was using this function's return value to help do that in the original implementation.

geoknee replied:
In the new design, channels are cleaned up only when the blocks they reference become safe, or when the entire state of the channel manager is reset. It is very much intentional to not garbage collect channels only when they are fully submitted+confirmed, because this is not guaranteed to preserve the ordering of the channels in the channel queue (see design doc linked above).

vdamle commented on op-batcher/readme.md:
```diff
@@ -0,0 +1,40 @@
+# op-batcher
```
Thank you so much for adding this and the architecture diagram @geoknee - super helpful! 🙌🏼 🎉

vdamle commented on op-batcher/batcher/channel_builder.go:
```diff
@@ -186,7 +192,7 @@ func (c *ChannelBuilder) AddBlock(block *types.Block) (*derive.L1BlockInfo, erro
 	}
 
-	c.blocks = append(c.blocks, block)
+	c.blocks.Enqueue(block)
```
Another question - why are the `blocks` & `frames` queues not protected from concurrent access (i understand this is the way the code was before as well)? I see that `op-service/queue` does not wrap append/read operations with an `RWMutex` either.

geoknee replied:
You are correct that `op-service/queue` is not designed to be inherently concurrency safe. The concurrency control is managed by the surrounding code where necessary.

The only async code in use at the moment (i.e. go routines) is in the `driver`, which protects the entire state (including all `Channels` and their `ChannelBuilders`) with a mutex. So I don't have concerns about concurrency, although it might be a good idea to run the race detector on some tests to see if that finds anything.

vdamle replied:
Good to close if this is not a concern. Just stood out to be, hence asked.

vdamle commented on op-batcher/batcher/channel_builder.go:
```diff
@@ -398,46 +404,51 @@ func (c *ChannelBuilder) Close() {
 	return c.numFrames
 }
 
 // HasFrame returns whether there's any available frame. If true, it can be
```
```suggestion
// HasFrame returns whether there's any pending frame. If true, it can be
```

vdamle commented on op-batcher/batcher/channel_builder.go:
```diff
@@ -398,46 +404,51 @@ func (c *ChannelBuilder) Close() {
 //
 // Call OutputFrames before to create new frames from the channel out
 // compression pipeline.
 func (c *ChannelBuilder) HasFrame() bool {
```
Not a strong opinion, IMO `HasPendingFrame()` feels more intuitive & accurate.

geoknee replied:
I went with your suggestion in the end.

vdamle commented on op-batcher/batcher/channel_builder.go:
```diff
@@ -398,46 +404,51 @@ func (c *ChannelBuilder) Close() {
 }
 
 // PendingFrames returns the number of pending frames in the frames queue.
 // It is larger zero iff HasFrame() returns true.
```
```suggestion
// It is larger than zero iff HasFrame() returns true.
```

vdamle commented on op-batcher/batcher/channel_builder.go:
```diff
@@ -398,46 +404,51 @@ func (c *ChannelBuilder) Close() {
 func (c *ChannelBuilder) NextFrame() frameData {
-	if len(c.frames) == 0 {
+	if len(c.frames) <= c.frameCursor {
 		panic("no next frame")
```
Probably an enhancement (maybe we should write up an issue) -- this (and elsewhere there are panics) feels more like an error that should be handled gracefully and logged/metric pushed as opposed to crashing the program? Or is there a reason why we want to `panic`?

geoknee replied:
Interesting one. I agree that this could probably be reworked to remove the need for a panic. It's not a great devex to say "you must do this other thing before calling this method or it will panic". The two methods could be combined most likely.

geoknee replied:
Filed an issue, could be a fast follow: https://github.com/ethereum-optimism/optimism/issues/12742

vdamle commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -70,14 +71,18 @@ func (s *channelManager) Clear(l1OriginLastClosedChannel eth.BlockID) {
 	defer s.mu.Unlock()
 	s.log.Trace("clearing channel manager state")
 	s.blocks.Clear()
+	s.blockCursor = 0
```
`blockCursor` should already have zero value by default, this is not needed

geoknee replied:
I agree that is the default value, but it may have changed during the course of the program. To be clear, this is not a constructor but a resetter.

vdamle commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -120,23 +115,23 @@ func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
+func (s *channelManager) rewindToBlockWithHash(blockHash common.Hash) {
+	for i, b := range s.blocks {
+		if b.Hash() == blockHash {
+			s.blockCursor = i
```
similar to frames, do we need to guard against the block cursor moving forward?

geoknee replied:
Good catch. Yes I think so, if for no other reason than this being true to the notion of "rewinding".

vdamle commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -120,23 +115,23 @@ func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
+// to point at the first block added to the provided channel,
+// and clears the channelQueue and currentChannel.
+func (s *channelManager) handleChannelTimeout(c *channel) {
+	blockHash := c.channelBuilder.blocks[0].Hash()
```
Should we check that the the blocks queue is non-empty here? or is the fact that we hit a channel timeout mean that there should have been blocks in the first place?

geoknee replied:
Good call out. 

By design, blocks are only ever enqueued to the channel builder. So the only way for there to be no blocks is if none were ever added. For a channel to timeout, at least one frame must have been submitted, which implies at least one block must have been added to the channel. I think a panic from trying to access a nil slice is next to impossible -- it could happen in an extreme fluke where a different batcher (or previous instance of this batcher) had a colliding channel id, and this batcher happened to have created a channel but not added any blocks to it yet (I'm not sure that can happen before this function runs anyway). 

IMO we need can rely on the uniqueness of channel ids here. If that fails, we probably have bigger problems.

geoknee replied:
Although I see below my own comment: 

> // This is usually true, but there is an edge case
		// where a channel timed out before any blocks got added.
		
So perhaps it is worth putting the check in here, and just having it be a no-op if there are no blocks.

vdamle commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -277,7 +267,7 @@ func (s *channelManager) ensureChannelWithSpace(l1Head eth.BlockID) error {
 		"l1Head", l1Head,
 		"l1OriginLastClosedChannel", s.l1OriginLastClosedChannel,
-		"blocks_pending", s.blocks.Len(),
+		"blocks_pending", s.blocks.Len()-s.blockCursor,
```
```suggestion
		"blocks_pending", s.pendingBlocks(),
```

vdamle commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -426,82 +416,93 @@ func l2BlockRefFromBlockAndL1Info(block *types.Block, l1info *derive.L1BlockInfo
-		}
+	if len(channelToDiscard.channelBuilder.blocks) > 0 {
+		// This is usually true, but there is an edge case
+		// where a channel timed out before any blocks got added.
```
Should we add a log for the edge case ( `else { log }`)?

geoknee replied:
Yes we could put a debug level log in for "discarded channel had no blocks").

geoknee committed:
typo

geoknee committed:
rewindToBlockWithHash never moves cursor forward

geoknee committed:
use s.pendingBlocks()

geoknee committed:
add log line

geoknee committed:
check there are blocks when handling timeout

geoknee committed:
rename HasFrame() to HasPendingFrame()

geoknee committed:
Merge remote-tracking branch 'origin/develop' into gk/batcher-cursors-rebased

geoknee committed:
fixup test

geoknee committed:
improve readme

geoknee committed:
link to open issues by tag

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -428,65 +438,36 @@ func (l *BatchSubmitter) loop() {
-			}
-			publishAndWait()
-			l.Log.Info("Finished publishing all remaining channel data")
 			return
```
still worth adding an info log that the main loop is returning.

geoknee replied:
[e2d6811](https://github.com/ethereum-optimism/optimism/pull/12390/commits/e2d68113bafcd8711f19c6088c2b01269c2fb9e5)

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -129,23 +124,34 @@ func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
+// rewindToBlockWithHash updates the blockCursor to point at
+// the block with the supplied hash, only if that block exists
+// in the block queue and the blockCursor is ahead of it.
+func (s *channelManager) rewindToBlockWithHash(blockHash common.Hash) {
```
We could also pass a `eth.BlockID` (hash and number) and then calculate the index at which the block must be, and sanity check the hash. If the hash is then incorrect, something must be wrong (reorg?).

geoknee replied:
[bb3c776](https://github.com/ethereum-optimism/optimism/pull/12390/commits/bb3c776e8513e8c2056fb4b4374bed96e721c993)

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -129,23 +124,34 @@ func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
+		s.log.Debug("channelManager.handleChannelTimeout: channel had no blocks")
 	}
-	s.channelQueue = append(s.channelQueue[:index], s.channelQueue[index+1:]...)
+	s.channelQueue = s.channelQueue[:0]
```
It seems inconsistent to rewind to the starting block of the channel, but then remove all channels in the queue. This only works if a timing out channel is actually always the first=oldest channel in the queue, _which probably is the case_. We should check that this is consistent.

geoknee replied:
[eda390d](https://github.com/ethereum-optimism/optimism/pull/12390/commits/eda390df6ead5b25332724507151477605e69c45)

sebastianst commented on op-batcher/batcher/driver.go:
```diff
@@ -428,65 +438,36 @@ func (l *BatchSubmitter) loop() {
 				l.clearState(l.shutdownCtx)
 				continue
 			}
 			l.publishStateToL1(queue, receiptsCh, daGroup)
```
For a follow up, we need to fix the `txmgr.Queue` to use `TxMgr.SendAsync`, or just use `SendAsync` directly, because we need to lock in the nonce at the point we submit the transaction. This should also have been fixed before Holocene, but with the strict ordering requirements, this becomes more important.

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -438,82 +439,95 @@ func l2BlockRefFromBlockAndL1Info(block *types.Block, l1info *derive.L1BlockInfo
+		// This could happen if there was an L1 reorg.
+		// We should restart work from the new safe head,
+		// and therefore prune all the blocks.
+		s.Clear(eth.BlockID{})
```
```suggestion
		s.Clear(newSafeHead.L1Origin.Number)
```

geoknee replied:
[e1e8beb](https://github.com/ethereum-optimism/optimism/pull/12390/commits/e1e8bebe383c102d7542c308976ec4d1c329893a)

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -438,82 +439,95 @@ func l2BlockRefFromBlockAndL1Info(block *types.Block, l1info *derive.L1BlockInfo
+			"newSafeBlock", newSafeHead)
+		// We should restart work from the new safe head,
+		// and therefore prune all the blocks.
+		s.Clear(eth.BlockID{})
```
We want to pass `newSafeHead.L1Origin.Number`

geoknee replied:
[e1e8beb](https://github.com/ethereum-optimism/optimism/pull/12390/commits/e1e8bebe383c102d7542c308976ec4d1c329893a)

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -438,82 +439,95 @@ func l2BlockRefFromBlockAndL1Info(block *types.Block, l1info *derive.L1BlockInfo
+	if numBlocksToDequeue > uint64(s.blocks.Len()) {
+		// This could happen if the batcher restarted.
+		// A previous batcher instance had the blocks in memory
+		// But now they are gone.
```
yes, and it's also that we derived the safe head from channels sent by a previous instance which already got more data onchain?

geoknee replied:
Clarified here [b7e87d3](https://github.com/ethereum-optimism/optimism/pull/12390/commits/b7e87d341d775b853cb9c51b7856fc36dc7e05ba)

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -438,82 +439,95 @@ func l2BlockRefFromBlockAndL1Info(block *types.Block, l1info *derive.L1BlockInfo
-		blocksToRequeue = append(blocksToRequeue, channel.channelBuilder.Blocks()...)
+	if s.blocks[numBlocksToDequeue-1].Hash() != newSafeHead.Hash {
+		ref, _ := derive.L2BlockToBlockRef(s.rollupCfg, s.blocks[numBlocksToDequeue-1])
+		s.log.Error("safe chain reorg",
```
warn-log

geoknee replied:
[a60ffe2](https://github.com/ethereum-optimism/optimism/pull/12390/commits/a60ffe2ab80a983de02d98988b1fc4cd0dff3e21)

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -438,82 +439,95 @@ func l2BlockRefFromBlockAndL1Info(block *types.Block, l1info *derive.L1BlockInfo
+		// This could happen if the batcher restarted.
+		// A previous batcher instance had the blocks in memory
+		// But now they are gone.
+		numBlocksToDequeue = uint64(s.blocks.Len())
```
instead, we should warn-log, `Clear(newSafeHead.L1Origin.Number)` and return.

geoknee replied:
[a60ffe2](https://github.com/ethereum-optimism/optimism/pull/12390/commits/a60ffe2ab80a983de02d98988b1fc4cd0dff3e21)

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -438,82 +439,95 @@ func l2BlockRefFromBlockAndL1Info(block *types.Block, l1info *derive.L1BlockInfo
-	for _, b := range blocksToRequeue {
-		s.metr.RecordL2BlockInPendingQueue(b)
+	if s.blockCursor < 0 {
+		panic("blockCursor got into bad state")
```
```suggestion
		panic("negative blockCursor")
```

geoknee replied:
[da89ebc](https://github.com/ethereum-optimism/optimism/pull/12390/commits/da89ebcecf2105d3984832abe762bc22e7207662)

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -438,82 +439,95 @@ func l2BlockRefFromBlockAndL1Info(block *types.Block, l1info *derive.L1BlockInfo
-		if err := s.outputFrames(); err != nil {
-			return fmt.Errorf("outputting frames during close: %w", err)
-		}
+	if newSafeHead.Number+1 < oldestBlock.NumberU64() {
```
is this covered in the tests?

geoknee replied:
It is now [71065c6](https://github.com/ethereum-optimism/optimism/pull/12390/commits/71065c6f83c44d7c38248e5a3bc8558437119c54)

sebastianst commented on op-batcher/batcher/channel_manager_test.go:
```diff
@@ -637,36 +418,146 @@ func TestChannelManager_Requeue(t *testing.T) {
-	require.NoError(t, m.ensureChannelWithSpace(eth.BlockID{}))
-	require.NotEmpty(t, m.channelQueue)
-	require.NoError(t, m.processBlocks())
+func TestChannelManager_PruneBlocks(t *testing.T) {
```
- need to add case where safe head is older the oldest block in the queue (clears)
- could use a test table instead, start with same chain, set safe head, define expected outcome

geoknee replied:
I added that case here [71065c6](https://github.com/ethereum-optimism/optimism/pull/12390/commits/71065c6f83c44d7c38248e5a3bc8558437119c54).

sebastianst commented on op-batcher/batcher/channel_manager_test.go:
```diff
@@ -637,36 +418,146 @@ func TestChannelManager_Requeue(t *testing.T) {
+	d := derivetest.RandomL2BlockWithChainId(rng, numTx, defaultTestRollupConfig.L2ChainID)
+	d = d.WithSeal(&types.Header{Number: big.NewInt(3)})
+	e := derivetest.RandomL2BlockWithChainId(rng, numTx, defaultTestRollupConfig.L2ChainID)
+	e = e.WithSeal(&types.Header{Number: big.NewInt(4)})
```
I'd call them `a0`, `a1`, `b2`, ... etc, depending on which channel they go in

geoknee replied:
[fe7eb67](https://github.com/ethereum-optimism/optimism/pull/12390/commits/fe7eb671d92fe5c64d8cfe91d3f12848b949f9f8)

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -129,23 +124,34 @@ func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
+// rewindToBlockWithHash updates the blockCursor to point at
+// the block with the supplied hash, only if that block exists
+// in the block queue and the blockCursor is ahead of it.
+func (s *channelManager) rewindToBlockWithHash(blockHash common.Hash) {
```
we could pass in a `BlockID` and then use the block number to calculate the cursor, instead of doing a loop, and then use the hash just to confirm whether we're on the same chain.

geoknee replied:
[c9635d7](https://github.com/ethereum-optimism/optimism/pull/12390/commits/c9635d7e71d753a4ee1c72a5fa4c32774a2de0dd)

sebastianst commented on op-batcher/batcher/channel_builder.go:
```diff
@@ -75,7 +76,12 @@ type ChannelBuilder struct {
 	oldestL2 eth.BlockID
 	// frames data queue, to be send as txs
-	frames []frameData
+	frames queue.Queue[frameData]
+	// frameCursor tracks which frames in the queue were submitted
+	// frames[frameCursor] is the next unsubmitted (pending) frame
+	// frameCursor = len(frames) is reserved for when
+	// there are no pending (next unsubmitted) frames
+	frameCursor int
```
We may consider adding a `CursorQueue` abstraction that is a `Queue` with a cursor and then use it for blocks and frames. This way, we can better unit test the invariants between the queue and cursor.

geoknee replied:
I like this idea but would prefer to tackle it in a followup, just to keep the diff on this PR smaller.

sebastianst replied:
Tracking at https://github.com/ethereum-optimism/optimism/issues/12926

sebastianst commented on op-batcher/batcher/channel_builder.go:
```diff
@@ -402,46 +408,51 @@ func (c *ChannelBuilder) Close() {
-func (c *ChannelBuilder) HasFrame() bool {
-	return len(c.frames) > 0
+func (c *ChannelBuilder) HasPendingFrame() bool {
+	return c.frames.Len() > 0 && c.frameCursor < c.frames.Len()
```
```suggestion
	return c.frameCursor < c.frames.Len()
```
should work, or it would be inconsistent, as the cursor can only be smaller or equal to the length

geoknee replied:
[81aa115](https://github.com/ethereum-optimism/optimism/pull/12390/commits/81aa1151734fc4cabd0da236725a6ac1490f7c30)

sebastianst commented on op-batcher/batcher/channel_builder.go:
```diff
@@ -402,46 +408,51 @@ func (c *ChannelBuilder) Close() {
+// RewindFrameCursor moves the frameCursor to point at the supplied frame
+// only if it is ahead of it.
+// Panics if the frame is not in this channel.
+func (c *ChannelBuilder) RewindFrameCursor(frame frameData) {
```
we should check if there's anything we wanna cover here that isn't covered yet.

geoknee replied:
[0ba74df](https://github.com/ethereum-optimism/optimism/pull/12390/commits/0ba74dfafe5779873dd7bcb6140b662037b59f75) and [23da037](https://github.com/ethereum-optimism/optimism/pull/12390/commits/23da0378aa35b712d312cd10f12ce5fdc2956ba8)

sebastianst reviewed (commented):
part 1

geoknee committed:
Merge remote-tracking branch 'origin/develop' into gk/batcher-cursors-rebased

geoknee commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -97,34 +102,21 @@ func (s *channelManager) TxFailed(_id txID) {
-			if len(blocksToRequeue) > 0 {
-				s.blocks.Prepend(blocksToRequeue...)
-			}
-			for _, b := range blocksToRequeue {
-				s.metr.RecordL2BlockInPendingQueue(b)
-			}
```
We need to make sure that we continue to track metrics properly, as this is something that has regressed in the past. Ideally, we add some tests to protect this.

geoknee replied:
[4cd021a](https://github.com/ethereum-optimism/optimism/pull/12390/commits/4cd021a06b1712cfb09d4447b199d8d3232f055c)

geoknee replied:
Added coverage for some metrics here [7a45ecd](https://github.com/ethereum-optimism/optimism/pull/12390/commits/7a45ecd795ff99ef3e99b23af0b3caf901b0a20d)

geoknee committed:
add log when main loop returns

geoknee committed:
pass blockID to rewindToBlock

and panic if block does not exist

geoknee committed:
don't remove all channels when a channel times out

keep older channels, it's possible that they also time out

geoknee commented on op-batcher/batcher/channel.go:
```diff
@@ -52,7 +52,7 @@ func (c *channel) TxFailed(id string) {
 		// this needs to be changed to iterate over all frames of the tx data
 		// and re-queue them.
-		c.channelBuilder.PushFrames(data.Frames()...)
+		c.channelBuilder.RewindFrameCursor(data.Frames()[0])
```
This is worth a comment to explain why we rewind to the first frame of the transaction which failed.

geoknee replied:
[753ad8e](https://github.com/ethereum-optimism/optimism/pull/12390/commits/753ad8ee795d5623ebd97824a30b081849b14add)

sebastianst commented on op-batcher/batcher/channel.go:
```diff
@@ -52,7 +52,7 @@ func (c *channel) TxFailed(id string) {
 		// this needs to be changed to iterate over all frames of the tx data
 		// and re-queue them.
-		c.channelBuilder.PushFrames(data.Frames()...)
+		c.channelBuilder.RewindFrameCursor(data.Frames()[0])
```
Worth a comment why we choose the first frame of the tx (because we order frames within a multi-frame tx).

geoknee replied:
[753ad8e](https://github.com/ethereum-optimism/optimism/pull/12390/commits/753ad8ee795d5623ebd97824a30b081849b14add)

geoknee commented on op-batcher/batcher/channel.go:
```diff
@@ -82,21 +80,20 @@ func (c *channel) TxConfirmed(id string, inclusionBlock eth.BlockID) (bool, []*t
 
+	if c.isFullySubmitted() {
+		c.metr.RecordChannelFullySubmitted(c.ID())
+		c.log.Info("Channel is fully submitted", "id", c.ID(), "min_inclusion_block", c.minInclusionBlock, "max_inclusion_block", c.maxInclusionBlock)
```
Should we augment the log to show whether the channel timed out or not?

geoknee committed:
use newSafeHead.L1Origin in Clear() when pruning blocks

geoknee committed:
clarify comment

geoknee committed:
use warn log level on safe chain reorg pruning, and unify handling for safe head above unsafe head

geoknee committed:
update panic message

geoknee committed:
extend test coverage and fix bug

geoknee committed:
rename test blocks

geoknee committed:
simplify HasPendingFrame() method

geoknee committed:
simplify implementation of RewindFrameCursor

geoknee committed:
activate dormant test

geoknee committed:
ensure pending_blocks_bytes_current metric is tracked properly

geoknee committed:
cover metrics behaviour in test

using new TestMetrics struct

geoknee committed:
extend test coverage to channelManager.handleChannelTimeout

geoknee committed:
add comment to TxFailed

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -133,23 +126,54 @@ func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
+		s.rewindToBlock(blockID)
+	} else {
+		s.log.Debug("channelManager.handleChannelTimeout: channel had no blocks")
+	}
```
Wondering if we want to move all of this into its own function, like `rewindToStart`, because it's also used in `Requeue`. Or maybe we use `Requeue` here? Since we also reset the `currentChannel` this seems to work.

geoknee replied:
I ended up combining / deduping the functions here [62bd303](https://github.com/ethereum-optimism/optimism/pull/12390/commits/62bd303fbc557d13adad41cc299450ab317cdb4f)

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -133,23 +126,54 @@ func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
+			if len(s.channelQueue) > i {
+				// keep older channels if there are any,
+				// in case we later learn they also timed out.
+				s.channelQueue = s.channelQueue[i+1:]
```
I'm confused, do we prepend new channels to the `channelQueue`? Looking at `ensureChannelWithSpace` we append new channels. So wouldn't you keep the _younger_ channels and throw away the older ones? 

Is this case tested? If not, we should add a test.

geoknee replied:
Yes sorry this was a bug. I added a failing test here [2d025e8](https://github.com/ethereum-optimism/optimism/pull/12390/commits/2d025e885ec57132a0c2efc292886bba769afdab) and then patched it here [74ed8b8](https://github.com/ethereum-optimism/optimism/pull/12390/commits/74ed8b81609a1bc7b1c143175765d6394c7056f6) and here [7b17b1c](https://github.com/ethereum-optimism/optimism/pull/12390/commits/7b17b1c693de904f2ec1cf36317db39348841e87)

sebastianst commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -133,23 +126,54 @@ func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
+			} else {
+				// if this is the oldest channel, clear the queue
+				s.channelQueue = s.channelQueue[:0]
+			}
```
So maybe this becomes
```suggestion
			s.channelQueue = s.channelQueue[:i+1]
```
?

geoknee replied:
Yes https://github.com/ethereum-optimism/optimism/pull/12390/commits/74ed8b81609a1bc7b1c143175765d6394c7056f6

geoknee replied:
Although note we need to exclude the actual channel which got invalidated https://github.com/ethereum-optimism/optimism/pull/12390/commits/7b17b1c693de904f2ec1cf36317db39348841e87.

sebastianst commented on op-batcher/batcher/channel_manager_test.go:
```diff
@@ -611,14 +394,32 @@ func TestChannelManager_TxData(t *testing.T) {
 
 // TestChannelManager_Requeue seeds the channel manager with blocks,
```
rename

geoknee replied:
[42727dc](https://github.com/ethereum-optimism/optimism/pull/12390/commits/42727dc950b12c118d372047599e29dfb434f23a)

sebastianst commented on op-batcher/readme.md:
```diff
@@ -0,0 +1,69 @@
+
+It depends directly on some code shared with the derivation pipeline, namely the [`ChannelOut`](../op-node/rollup/derive/channel_out.go) implementation(s). It also depends directly on the shared [txmgr](../op-service/txmgr/) module.
+
+## Testing
```
There's also a bunch of op-e2e tests that test some aspects of the batcher.

geoknee replied:
[869b745](https://github.com/ethereum-optimism/optimism/pull/12390/commits/869b745adec5078c53ffab69f047a75abdf9b7a7)

sebastianst commented on op-batcher/readme.md:
```diff
@@ -0,0 +1,69 @@
+
+Link to [open issues with the `op-batcher` tag](https://github.com/ethereum-optimism/optimism/issues?q=is%3Aopen+is%3Aissue+label%3AA-op-batcher).
+
+The batcher launches L1 transactions in parallel so that it can achieve higher performance, particularly in situations where there is a large backlog of data which needs to be posted. Sometimes, transactions can get stuck in the L1 mempool. The batcher does have functionality to clear these stuck transactions, but it is not completely reliable.
```
```suggestion
The batcher launches L1 transactions in parallel so that it can achieve higher throughput, particularly in situations where there is a large backlog of data which needs to be posted. Sometimes, transactions can get stuck in the L1 mempool. The batcher does have functionality to clear these stuck transactions, but it is not completely reliable.
```

geoknee replied:
[9ceff07](https://github.com/ethereum-optimism/optimism/pull/12390/commits/9ceff0754e149aadcb4015dd57345e6c337ba15e)

sebastianst reviewed (commented):
There's one confusion in `handleChannelTimeout` that we need to sort out. Otherwise lgtm, just minor comments and improvement proposals.

geoknee committed:
rename test fn

geoknee committed:
point to e2e tests in readme.

geoknee committed:
readme: performance -> throughput

geoknee committed:
improve channel_manager_test to assert old channels are not affected by requeue or timeout

geoknee committed:
fix handleChannelTimeout behaviour

We were trimming older channels and keeping new ones. We need to trim newer channels and keep old ones. Fixes associated test (see previous commit).

geoknee committed:
tighten up requirements for invalidating a channel

geoknee committed:
replace requeue with handleChannelInvalidated

geoknee committed:
Merge remote-tracking branch 'origin/develop' into gk/batcher-cursors-rebased

samlaf commented on op-batcher/readme.md:
```diff
@@ -0,0 +1,69 @@
+
+
+### Reorgs
+When an L2 unsafe reorg is detected, the batch submitter will reset its state, and wait for any in flight transactions to be ingested by the verifier nodes before starting work again.
```
What is an L2 unsafe reorg wrt:
<img width="194" alt="image" src="https://github.com/user-attachments/assets/75f91eb3-3a48-42b6-b3b1-9f3ec70d71fc">

Even safe L2 blocks (submitted on L1) could be reorged out by an L1 reorg right? Did you mean L1 unsafe (or safe) reorg here?

geoknee replied:
Yes safe L2 blocks can be reorged by an L1 reorg.

What I mean here by an L2 reorg, is when the unsafe blocks being pulled from the sequencer don't descend from the previous L2 blocks in the batcher's memory. Does that make sense?

samlaf replied:
Yep that makes sense! Could we rephrase to just say "ingested by the sequencer's op-node" here? Or are there cases the batcher is really waiting for nodeS?

And just to be sure, this kind of reorg can only happen when an L1 reorg happens right..?

geoknee replied:
Yes we can make that edit.

An unsafe reorg can happen when e.g. the batcher is down for a long time such that the sequencing window is breached https://docs.optimism.io/stack/rollup/derivation-pipeline.

samlaf replied:
On reading the code more here, this same logic would also be applied for safe reorgs right? So think this should say both unsafe and safe reorgs?

geoknee replied:
Yes good point. Could you send a PR for these readme changes?

Also, the logic for handling these reorgs is being refactored here #13060

samlaf replied:
Sounds good, I'll wait for that PR to land and then refactor my PR on top of it. Also I'll include the readme changes in my PR to force you to review it :D
